### ISSUE TITLE
06/28

### ISSUE THREAD

**Issue by youngsun807** [youngsun807/WONIK_KAIST_FTC#1]
### pip VS conda
- pip = 버전, 환경변수 manual로 설정해야 함
- conda = 환경에 맞게 깔 수 있음

# deep learning basics
- I.I.D를 따르지 않는 경우 : 현재 FTC 온도가 서로 앞뒤 스텝별로 영향을 주기 때문에 무시하고 사용 중
- inversion 관련 수식 사용 X -> 복잡도 높고, 역행렬 없을 수 있음
- linear + non linear 모델 합성해서 사용하면 복잡한 모델 활용 가능 -> non linear는 미분만 가능한 것들 모두 가능
- non linearity 없으면 계속해서 새로운 Linear 가설이 생성되서 복잡, 무쓸모가 됨
- automatic differentiation 코드로 미분 계산
- momentum = 기존 방향대로 진행 -> 감마만큼 움직임 모멘텀을 주고 계산
- 현재 움직일 때 과거 움직임을 가지고 계산함(exponential average)
- adagrad = adapted gradient(크게 움질일 때는 크게, 작게 움직일 때는 작게 경사도 부여) -> 많이 움직였으면 충분히 최적화 된 것이라 가정 -> 많이 움직였음 stepsize 줄여나가기
- 감마/루트(Gt + 입실론) 값이 Gt가 커지게 되면 많이 학습하게 되면 값의 변화가 없고 너무 커져서 더 가야되지만 가지 못할 때 RMSprop 기법 사용
- 메모리 내 경사 하강법을 다 계산할 수 없어 샘플링해서 조금씩 계산 -> 전체 dataset에서 랜덤하게 뽑아내서 stochastic이라고 이야기한다 -> 샘플의 에러들의 평균을 내서 경사 하강법 진행함 -> 성능 훨씬 좋음
- ReLU -> 0인 구간에서 빠져나오지를 못함 -> 음수쪽에 기울기를 준다(leaky ReLU, 경사도도 설정 가능)
- 여러 libarary를 이용하여 가장 잘 나오는 initialization 기법을 알아서 찾을 수 있음
- batch normalization = SGD 한 샘플에 대해서 정규화 진행 -> 튀는 것 정규화 -> 배치 평균, 분산 게산 -> 감마 = mean, 베타 = 분산과 같아짐
- batch norm. 했을 때 성능 나오는 step 횟수 감소됨 -> 배치 사이즈 1일 경우 시그마 계산 불가해서 이것에 대해 batch norm 사용 불가
- L1 regularization(네트워크 compression -> 0인 인자들에 대해 계산 배제 가능해서 사용) = theta1 = 절대값 -> u = penalty에 쓰이는 것 = 벡터 크기 구하기
- L2 regularizaion = theta2 = 유클리디안 거리 계산값 = 각 값들의 거리 계산 = 루트(각 파라미터 값 제곱)
- dropout 기법(2^n) = input 중 일부 랜덤하게 0으로 만듦dropout -> 일종의 뉴런들이 영향력이 없다고 판단 -> 일종의 앙상블 -> inter correlation을 없애줄수도 있다는 장점도 있음
- droupout 시 테스트할 때 일단 input 전체 다 켜놓고 켜질 확률을 곱해주는 형식
- 양상블 = 여러개가 합쳐져서 하나의 모델이 되는 것으로 여러 기법 합치는 것 -> 각 모델 결과값 평균 내서 사용

# 실습
- numpy, pytorch, cuda random seed fix -> 재현성 위해
- dataset, dataloader, fitting 알고리즘이 성능, 시간에 중요한 요소? 인듯
